### PR TITLE
1992 mpi support for extraction

### PR DESCRIPTION
**Pull request by hiker**
Adds (limited) support for using kernel extraction with MPI. The idea is that each process writes its own output file. Kernel extraction cannot be done over nodes that do halo exchange or other communication. The driver can then run over each individual output file (from each rank)

### CODE REVIEW

File: src/psyclone/tests/domain/gocean/transformations/gocean_extract_test.py [hunk 1]
```diff
@@ -670,4 +657,5 @@ def test_change_prefix(monkeypatch):
     driver_name = "driver-main-update.f90"
     with open(str(driver_name), "r", encoding="utf-8") as driver_file:
         driver_code = driver_file.read()
-    assert "call NEW_psy_data%OpenRead('main', 'update')" in driver_code
+    assert ("call NEW_psy_data%OpenReadModuleRegion('main', 'update')"
```
Comment (arporter) [inline src/psyclone/tests/domain/gocean/transformations/gocean_extract_test.py:660]: Would it make sense to add a test when DM is enabled in this file too or is it a base-class thing? Actually, even if it is a test would be good I think.

File: src/psyclone/tests/domain/lfric/lfric_extract_driver_creator_test.py [hunk 2]
```diff
@@ -270,7 +270,11 @@ def test_lfric_driver_simple_test():
     with open(filename, "r", encoding='utf-8') as my_file:
         driver = my_file.read()
 
-    for line in ["call extract_psy_data%OpenRead('field', 'test')",
+    for line in ["if (ALLOCATED(psydata_filename)) then",
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_extract_driver_creator_test.py:273]: Ah, so we generate the same code irrespective of whether DM is enabled or not - it's a compile-time thing?

**Review by arporter — CHANGES_REQUESTED** [review]
All looks good Joerg, thanks. Just needs a bit of tidying here and there. I can see you've previously run the integration tests so I won't do that this time around.
Updated UG builds fine.
Ref. guide builds with no new warnings.

File: examples/gocean/eg5/extract/test.X90 [hunk 3]
```diff
@@ -1,4 +1,8 @@
 Program test
```
[the remark this answers is not in the document]
Comment (hiker) [reply examples/gocean/eg5/extract/test.X90:1]: Done.

File: examples/lfric/eg17/full_example_extract/main.X90 [hunk 4]
```diff
@@ -33,6 +33,10 @@
 
 program main
```
[the remark this answers is not in the document]
Comment (hiker) [reply examples/lfric/eg17/full_example_extract/main.X90:43]: Done.

File: lib/extract/netcdf/extract_netcdf_base.jinja [hunk 5]
```diff
@@ -203,6 +203,9 @@ contains
     subroutine PreStart(this, module_name, region_name, num_pre_vars, &
                         num_post_vars)
 
+#ifndef NO_MPI
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/extract_netcdf_base.jinja:206]: The reason it relies on preprocessing is that the files will get copied into the LFRic build tree (which is by far the easiest option to get them properly integrated into the right version of the LFRic infrastructure). So the source files need to work with MPI and without.

File: lib/extract/standalone/read_kernel_data_mod.jinja [hunk 6]
```diff
@@ -142,7 +144,30 @@ contains
...
-    end subroutine OpenRead
+    end subroutine OpenReadModuleRegion
+
+    ! -------------------------------------------------------------------------
+    !> @brief This subroutine is called to open a binary file for reading. The
+    !! filename is based on the module and kernel name. This is used by a
+    !! driver program that will read a binary file previously created by the
+    !! PSyData API.
+    !! @param[in,out] this The instance of the ReadKernelDataType.
+    !! @param[in] module_name The name of the module of the instrumented
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/read_kernel_data_mod.jinja:156]: Oops, yes, here and in several other places :(

File: lib/extract/standalone/read_kernel_data_mod.f90 [hunk 7]
```diff
@@ -160,7 +162,30 @@ subroutine OpenRead(this, module_name, region_name)
...
 
-    end subroutine OpenRead
+    end subroutine OpenReadModuleRegion
+
+    ! -------------------------------------------------------------------------
+    !> @brief This subroutine is called to open a binary file for reading. The
+    !! filename is based on the module and kernel name. This is used by a
+    !! driver program that will read a binary file previously created by the
+    !! PSyData API.
+    !! @param[in,out] this The instance of the ReadKernelDataType.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/read_kernel_data_mod.f90:173]: Indeed (actually, in the jinja file).

File: src/psyclone/tests/domain/gocean/transformations/gocean_extract_test.py (hunk 1 above)
Comment (hiker) [reply src/psyclone/tests/domain/gocean/transformations/gocean_extract_test.py:660]: I've added this.

File: src/psyclone/tests/domain/lfric/lfric_extract_driver_creator_test.py (hunk 2 above)
Comment (hiker) [reply src/psyclone/tests/domain/lfric/lfric_extract_driver_creator_test.py:273]: Yes, nothing changes in the driver - we just potentially read from different input file(s). We can discuss at some stage if we want 'more' MPI support (though that's difficult, we would need to store info about domain decomposition, ... tbh ... let's not go there unless we have a real urgent need :)  )

File: doc/user_guide/psyke.rst [hunk 8]
```diff
@@ -153,11 +153,16 @@ are used or not.
...
-As noted in the :ref:`PSyKAl Introduction <introduction_to_psykal>` section,
-PSyKAl can support distributed memory. However, since the generated PSy-layer
-code with DM enabled contains infrastructure calls (e.g. checks for runtime
-status of field halos, halo exchanges etc.), code extraction is not allowed
-when distributed memory is enabled.
+Kernel extraction for distributed memory is supported in as much as each
+process will write its own output file by adding its rank to the output
+file name. So each kernel and each rank will produce one file. It is possible
+to extract several consecutive kernels, but there must be no halo exchange
+call between the kernels. The extraction transformation will test for this
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/user_guide/psyke.rst]: Done.

File: doc/user_guide/psyke.rst [hunk 9]
```diff
@@ -153,11 +153,16 @@ are used or not.
...
+and raise an exception if this should happen.
+The compiled driver program accepts the name of the extracted kernel file as command
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/user_guide/psyke.rst]: Done.

File: doc/user_guide/psyke.rst [hunk 10]
```diff
@@ -381,6 +388,19 @@ and in
...
+All versions of the extraction libraries can be compiled with MPI
+support by setting the variable ``MPI=yes``:
+
+.. code-block:: shell
+
+  make MPI=yes ...
+
+The only difference is that the output files will now have the process
+rank in the name. Issue #2382 keeps track of improving the driver program
+to be able to read several data files, at this stage the driver will only
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/user_guide/psyke.rst]: I've rephrased it, I don't think we need a reference to 2382 here, it's a feature enhancement that we might want to consider at some stage.

File: lib/extract/netcdf/read_kernel_data_mod.f90 [hunk 11]
```diff
@@ -184,7 +185,30 @@ subroutine OpenRead(this, module_name, region_name)
         retval = CheckError(nf90_open(module_name//"-"//region_name//".nc", &
                                         NF90_NOWRITE, this%ncid))
 
-    end subroutine OpenRead
+    end subroutine OpenReadModuleRegion
+
+    ! -------------------------------------------------------------------------
+    !> @brief This subroutine is called to open a NetCDF file for reading. The
+    !! filename is based on the module and kernel name. This is used by a
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/read_kernel_data_mod.f90]: Indeed. Fixed.

File: lib/extract/netcdf/read_kernel_data_mod.jinja [hunk 12]
```diff
@@ -166,7 +167,30 @@ contains
         retval = CheckError(nf90_open(module_name//"-"//region_name//".nc", &
                                         NF90_NOWRITE, this%ncid))
 
-    end subroutine OpenRead
+    end subroutine OpenReadModuleRegion
+
+    ! -------------------------------------------------------------------------
+    !> @brief This subroutine is called to open a NetCDF file for reading. The
+    !! filename is based on the module and kernel name. This is used by a
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/read_kernel_data_mod.jinja]: Yes, the source of all the confusion.

File: lib/extract/netcdf/read_kernel_data_mod.f90 [hunk 13]
```diff
@@ -184,7 +185,30 @@ subroutine OpenRead(this, module_name, region_name)
...
+    !! driver program that will read a NetCDF file previously created by the
+    !! PSyData API.
+    !! @param[in,out] this The instance of the ReadKernelDataType.
+    !! @param[in] module_name The name of the module of the instrumented
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/read_kernel_data_mod.f90]: Done (it's annoying, why does github not display the newly changed text here, that would make it easier for me to verify that I indeed rebuild everything).

File: lib/extract/standalone/dl_esm_inf/README.md [hunk 14]
```diff
@@ -43,6 +43,17 @@ environment variables ``$F90`` and ``$F90FLAGS`` can be set to point to the
...
 default to ``gfortran`` and the empty string.
 
+The library can be compiled with MPI support, which will add the process rank
+to each output filename, by setting ``MPI=yes``:
+
+```shell
+MPI=yes make
+```
+
+The build environment will provide ``NO_MPI`` as pre-processor definition when
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/dl_esm_inf/README.md]: Done.

File: lib/extract/standalone/lfric/README.md [hunk 15]
```diff
@@ -39,6 +39,21 @@ environment variables ``$F90`` and ``$F90FLAGS`` can be set to point to the
...
 default to ``gfortran`` and the empty string.
 
+The library can be compiled with MPI support, which will add the process rank
+to each output filename, by setting ``MPI=yes``:
+
+```shell
+MPI=yes make
+```
+
+It uses ``NO_MPI`` as preprocessor directive, which is what the LFRic build
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/README.md]: Done - yes, that name is so annoying :)

File: src/psyclone/domain/lfric/lfric_extract_driver_creator.py [hunk 16]
```diff
@@ -733,6 +735,88 @@ def _add_precision_symbols(symbol_table):
...
+        alternative filename (to the default one that is hard-coded by
+        the created driver) can be specified, which allows the driver to
+        be used with different files, e.g. several dumps from one run, and/or
+        a separate file from each process. It will also add the code to
+        open the input file using PSyclone's read_kernel_data module.
+
+        :param program: The driver PSyIR.
+        :type program: :py:class:`psyclone.psyir.nodes.Routine`
+        :param psy_data_var: the symbol of the PSyDataExtraction type.
+        :type psy_data_var: :py:class:`psyclone.psyir.symbols.Symbol`]]
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/domain/lfric/lfric_extract_driver_creator.py]: Removed.

**Comment by hiker** [comment]
OK, I think I have addressed everything. Not sure if you want to re-trigger CI, I think I only changed comments and documentation.

File: lib/extract/standalone/read_kernel_data_mod.f90 (hunk 7 above)
[the remark this answers is not in the document]
Comment (arporter) [reply lib/extract/standalone/read_kernel_data_mod.f90:173]: You've missed the `@param[in]...` bits below?

File: lib/extract/standalone/read_kernel_data_mod.jinja (hunk 6 above)
[the remark this answers is not in the document]
Comment (arporter) [reply lib/extract/standalone/read_kernel_data_mod.jinja:156]: Still here?

**Review by arporter — CHANGES_REQUESTED** [review]
Just a missed docstring update in a jinja file and then this is done.
I've fired off the integration tests just to be on the safe side.

File: lib/extract/standalone/read_kernel_data_mod.jinja (hunk 6 above)
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/read_kernel_data_mod.jinja]: Oops, sorry :(

**Comment by hiker** [comment]
Ready.

**Review by arporter — APPROVED** [review]
Requested changes have all been made, thanks.
Integration tests were all green.
Will proceed to merge.

**Pull request merged**